Sources/NIOPosix/BSDSocketAPICommon.swift

```diff
@@ -127,6 +127,333 @@ extension NIOBSDSocket.Option {
         NIOBSDSocket.Option(rawValue: Posix.IPV6_RECVPKTINFO)
 }
 
+extension NIOBSDSocket {
+    /// In the Internet Protocol version 4 (IPv4) [RFC791] there is a field
+    /// called "Protocol" to identify the next level protocol.  This is an 8
+    /// bit field.  In Internet Protocol version 6 (IPv6) [RFC8200], this field
+    /// is called the "Next Header" field.
+    public struct ProtocolSubtype: RawRepresentable {
+        public typealias RawValue = CInt
+        public var rawValue: RawValue
+        
+        @inlinable public init(rawValue: RawValue) {
+            self.rawValue = rawValue
+        }
+    }
+}
+
+// Generated with:
+//
+// import Foundation
+//
+// let protocols = try String(contentsOf: URL(filePath: "/etc/protocols"))
+//
+// let regex = #/^([^#].+)\t+(\d+)\t+([\w\-\+/]+)\t*#*\ *(.+)?$/#.anchorsMatchLineEndings()
+// for match in protocols.matches(of: regex) {
+//     let (_, name, number, _, description) = match.output
+//
+//     guard let number = CInt(number) else { continue }
+//
+//     let identifier = name
+//         // hyphens to camel case
+//         .replacing(#/-([a-z1-9])/#) { match in
+//             match.output.1.uppercased()
+//         }
+//         // 3pc has a number at the front which and therefore an illegal Swift identifier.
+//         // This is the only name with a number in the front and we can just hard code a different identifier
+//         .replacing("3pc", with: "thirdpc")
+//         // tp++ -> tpPlusPlus
+//         .replacing("+", with: "Plus")
+//         // a/n -> an
+//         .replacing(#/[\./]/#, with: "")
+//
+//
+//     if let description {
+//         print("/// \(description)")
+//     }
+//     print("public static let \(identifier) = Self(rawValue: \(number))")
+// }
+extension NIOBSDSocket.ProtocolSubtype {
+    /// internet protocol, pseudo protocol number
+    public static let ip = Self(rawValue: 0)
```

Comment:
These protocol subtypes are not necessarily stable from OS release to OS release. As much as possible, can we use the symbolic names?

Reply:
Where can I find the symbolic names?

Reply:
Some of these have that effect, but some do not. The protocol subtype field has uses outside of raw sockets.

In this case, these all match `IPPROTO_*` constants in the various libcs. I suspect `netinet/in.h` has most of them.

Reply:
I have made `NIOBSDSocket.ProtocolSubtype` internal and moved all known protocol types to a new type called `NIOIPProtocol` which is public. Should I still use the `IPPROTO_*` constants for those as well?

Reply:
I think if we can point to the IANA registry that's sufficient.

Reply:
Not sure I can follow. What do you mean by pointing to the IANA registry?

Reply:
IANA (the Internet Assigned Numbers Authority) holds registries for most extension points in IETF protocols. In the case of the IPv4 next-level protocol field or IPv6 next header field, that registry is [this one](https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml). You can skip most of that registry: I'd limit myself to assignments with a reference to an RFC or IANA. But that's the authoritative reference for the numbers, so we should reference it as the source for the numbers we choose.

Reply:
Done in https://github.com/apple/swift-nio/pull/2317/commits/30e96bba9f72346fd43e94bdb9f5b4e68398eb05